### DISCUSSION TITLE
Inconcistant checkbox behavior

### DISCUSSION THREAD

**Discussion by d3cline** [alpinejs/alpine#1890] [1. Help]
Say I got this template, with a checkbox in it, 
```
                <input
                  x-model="selected_mailboxes[item.checkIndex]"
                  type="checkbox"
                  name="checkbox_mailbox"
                  class="form-checkbox text-brandypunch rounded mx-4 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-emerald"
                  x-bind:data-id="item.id"
                  x-bind:data-name="item.name"
                  >
```

And the checkbox's index is kept server side to fix some shortcomings in alpine  https://github.com/alpinejs/alpine/discussions/1096

In alpine 2 this seemed to work, in alpine 3 however we get this 'empty slots' stuff

```
Proxy { <target>: (4) […], <handler>: {…} }
​
<target>: Array(4) [ <2 empty slots>, false, true ]
​
<handler>: Object { get: we(n, i, o), set: ar(r, n, i, o), deleteProperty: _n(e, t), … }
```

TLDR how to make alpine do this correctly? it is very common to need an iterated, paginated, list of objects with a checkbox, and as is template can't do it. The variable is able to track clicks, but not in a sane consistent way which make it easy to write code which does anything else.

**Comment by SimoTod** [comment]
What is the issue exactly?
The debug bit is a proxy object wrapping an array. It looks like you populated index 2 and 3 so position 0 and 1 are undefined.

**Reply by d3cline** [reply]
The live code is pretty complex. Basic examples all seem to work. The issue arises when we add features to a list of checkboxes, when the template is re-rendered based on some other action such as pagination or table sort. Lets say we have a JSON array with objects. It is fed into a filter function, and the return is used as the input for x-for. In our platform they can be anything, but lets say for this example they are certificates. They have object keys like name, etc. The ui has a filter function which returns a new object filtered. There is another part of the function to sort ascending or descending (screenshots) by rearrange the original array and returning the 'filtered' object Both pagination and sort use the same 'filter' function and return. Some kind of disconnect however is happening which applies 'true' to the other objects. Attached is an example with a single click. The first Image is 2 objects selected normally we are good up to this point. The second image is once I return the new object and the template redraws, showing now 4 true statements where only 2 should exist. 

![image](https://user-images.githubusercontent.com/77483/128101236-e1e6d485-e099-46fe-bbc5-8fecc6d4d4d8.png)
![image](https://user-images.githubusercontent.com/77483/128101252-50217c17-ece9-4a42-ba05-778de8b8d101.png)

I have attempted a few approaches to make this work, but it is unintuitive in alpine. I am actually removing the feature totally until it is more mature. But I would like to see some kind of working example if anyone else has one.
